### PR TITLE
Network regularization implementation + simulations

### PR DESCRIPTION
**Pull request by jjc2718**
This PR implements the graph Laplacian-based network regularization penalty described in https://doi.org/fk8n4b (and many subsequent papers using related ideas). The main implementation of this is in `utilities/pytorch_model.py`, and an example of how it is used can be found in `netreg_benchmark.py`.

To confirm this is functioning properly, I generated simulated data using the same method as in [this notebook](https://github.com/dirmeier/netReg/blob/master/examples/example.ipynb) (also described in more detail in the documentation for `simdata/simulate_networks.py`). Results of fitting network regularized and non-regularized models are shown in the notebook `simdata/simulation_analysis.ipynb` (spoiler: the network regularized models do indeed perform better for this simulation).

I also added some code to plot learning curves, which is often useful for debugging these models.

### CODE REVIEW

File: utilities/pytorch_model.py [hunk 1]
```diff
@@ -200,32 +258,81 @@ def torch_model(self, X_train, X_test, y_train, y_test, params,
...
+                ).detach()))
+                self.monitor_['test_loss'].append(float((
+                    criterion(y_pred_test, y_ts)
+                ).detach()))
+                # also save l1 loss
+                self.monitor_['l1_loss'].append(float((
+                    sum(torch.norm(param, 1)
+                            for name, param in model.named_parameters()
+                            if 'bias' not in name)
+                ).detach()))
```
Comment (ben-heil) [inline utilities/pytorch_model.py]: Calling `.item()` here my fill the role of both the `float` cast and the call to `.detach()`. The parenthesis nesting is a little unclear to me though, so I'm not sure

File: utilities/pytorch_model.py [hunk 2]
```diff
@@ -200,32 +258,81 @@ def torch_model(self, X_train, X_test, y_train, y_test, params,
...
+                # also save L2 network loss
+                network_weights = [param for name, param in model.named_parameters()
+                                         if 'bias' not in name][0]
+                network_weights = network_weights[:, self.network_features]
+                network_loss = torch.mm(
+                    w.view(1, -1),
+                    torch.sparse.mm(self.laplacian, w.view(-1, 1)))
+                self.monitor_['network_loss'].append(float((
+                    (network_loss).view(-1)[0]
+                ).detach()))
```
Comment (ben-heil) [inline utilities/pytorch_model.py]: You may have already considered this, but you could use a tensorboard [summarywriter](https://pytorch.org/docs/stable/tensorboard.html) to serialize information to a file instead of storing everything in a variable. If you do, you get visualization for free via tensorboard

File: simdata/simulation_analysis.ipynb [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
...p/GUjgYe/1HSD34K8lrplOo9RAMQ6driYiIiIi4i9uaEACtW7emdevWmR6bPHkyANWrV2f37t3ufHkRkStK2b6U5E3fUOSRMViKFMNR/xGSV03Fuedn/Co19nZ5IiIiIiI+S1/55TGjR79CdPSxXN1nZGTdXN3fv9WuXWuOHTvK0aNHGDPmVW+XIwWQrXQtcKWRvO4LAPwqNcYSVp7kX77CTDrr5erE25TDIpKTRYsW0aJFC+677z5mzpx52fM7d+7k4Ycfpk2bNvTu3Zv4+HgA9u3bR6dOnWjbti0dOnTg999/B+DIkSPUqlWLtm3b0rZtW3r06OHR48mLlMUivktNiDzm1183YZpmziv6gOjoYxw5ovPwxfMsRUOx12yFc/8GnH/vwDAs+Ed2xUw5T/KGrwFwnT9D2tnTXq5UvEE5LCLZOX78OOPHj2fWrFnMnz+fr776ij///DPTOqNGjaJv374sXLiQMmXKMGXKFACGDRtGz549WbBgAf3792fw4MEA7Nixg9atW7NgwQIWLFiQsX5BpiwW8V1uPR1D0gN0xoyp+Pv7c/DgAcqVK8/LL49i+fKlfP31F7hcJpUqVea//x3M7NlfcOJELIMG9aNly7acOhXH00/3ZcOGXxg2bDCLF6/AZrPRqVM73n33Y44ePcrEiW+RkpJCcHAwgwa9SMmSpejTpxdFiwZx4MA+Xn11TEYt27dvZdSoEbz11juULFnqivUuXryItWtXc+rUKU6ePEGjRo3p02cAhmEwY8Y0fvzxe9LSXNSv34CnnupLdPQxXnxxIGXLlmPPnj8oXrwEI0e+TtGiQcyd+xVLliwmKSkRPz8/RowYxS233JrxWhMnvsXRo0cYN+4Nzp8/S40atWnTJv02rX369OKpp/pSrdptbv35SMFlr9Gc1L1rSfp5BkXavYa1RCn8bruX1O1L8avUGDPxDE5dMNcnKIeVwyK5ae3atTRo0IDg4GAAmjVrxpIlS+jTp0/GOi6Xi3PnzgGQmJhIUFAQAO3bt6dx4/TT/ipVqsSxY+nf9G/fvp09e/bQtm1bgoKCGDp0KJUqVfLkYbmdslhZLHKBmhAesGPHNmbOnMMNN4TQu3c3vvlmDitXruDDDz/F4XDw0Ufv8cUXM+jW7QkWLJjLm29OJCUlhVdfHQ7A5s0bcTgc7Nmzm+DgYhQpEkBgYFFGjOjGy
```
Comment (ben-heil) [inline simdata/simulation_analysis.ipynb]: Yay! Do you know why the pytorch implementations seem to have a higher variation in model performance than those of sklearn when there are many features?

File: scripts/run_netreg.R [hunk 4]
```diff
@@ -0,0 +1,84 @@
+suppressPackageStartupMessages(library(argparse))
```
Comment (ben-heil) [inline scripts/run_netreg.R]: I feel like they missed an opportunity in not calling their package Rgparse

File: utilities/pytorch_model.py [hunk 5]
```diff
@@ -40,31 +45,67 @@ def __init__(self,
...
+        # if both network_file and network_penalty are provided, load network
+        # data and calculate graph Laplacian
+        if network_file is not None and params_map['network_penalty'] != 0:
+            assert network_features is not None, (
+                'list of features in network should be included')
+            self.network_features = network_features
+            import networkx as nx
+            G = nx.read_weighted_edgelist(network_file, delimiter='\t')
+            lt = nx.laplacian_matrix(G)
+            indices, values, shape = self._convert_csr_to_sparse_inputs(lt)
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply utilities/pytorch_model.py]: I'll add a note to the code, but in case you're curious: it's a format for storing sparse matrices in a memory-efficient way ([see here](https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_row_\(CSR,_CRS_or_Yale_format\)) for more details). PyTorch uses a different format, so I have to convert it here.

File: utilities/pytorch_model.py [hunk 6]
```diff
@@ -40,31 +45,67 @@ def __init__(self,
...
             torch.backends.cudnn.deterministic = True
             torch.backends.cudnn.benchmark = False
 
+        # if both network_file and network_penalty are provided, load network
+        # data and calculate graph Laplacian
+        if network_file is not None and params_map['network_penalty'] != 0:
+            assert network_features is not None, (
+                'list of features in network should be included')
+            self.network_features = network_features
+            import networkx as nx
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply utilities/pytorch_model.py]: Nope, I'll move it

File: utilities/pytorch_model.py [hunk 7]
```diff
@@ -168,28 +220,34 @@ def torch_model(self, X_train, X_test, y_train, y_test, params,
...
+                print('\n[0, 1]: {} (pos_weight={:.4f})'.format(train_count, pos_weight))
 
         if self.use_gpu:
             X_tr = torch.stack([torch.Tensor(x) for x in X_train]).cuda()
             X_ts = torch.stack([torch.Tensor(x) for x in X_test]).cuda()
             y_tr = torch.Tensor(y_train).view(-1, 1).cuda()
             y_ts = torch.Tensor(y_test).view(-1, 1).cuda()
-            pos_weight = torch.Tensor([pos_weight]).cuda()
+            if classify:
+                pos_weight = torch.Tensor([pos_weight]).cuda()
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply utilities/pytorch_model.py]: Good call! I do want to be able to choose how to train (sometimes things are faster on the CPU, if the overhead of moving data to the GPU isn't worth the number of computations), but I think what you're suggesting with the `device` variable should make things more concise.

File: utilities/pytorch_model.py (hunk 2 above)
Comment (jjc2718) [reply utilities/pytorch_model.py]: I did consider this, but I was trying to quickly debug a specific issue so I just threw together a simple solution. Longer-term I think using TensorBoard is probably more flexible though - created #18 to switch over in the future.

File: utilities/pytorch_model.py (hunk 1 above)
Comment (jjc2718) [reply utilities/pytorch_model.py]: That seems to work!

File: simdata/simulation_analysis.ipynb (hunk 3 above)
Comment (jjc2718) [reply simdata/simulation_analysis.ipynb]: I'm not sure. The only difference that comes to mind is that I'm doing minibatch GD in pytorch and (I think) sklearn uses single-sample SGD. Not sure why that would cause more variability in pytorch though, I'd actually expect the opposite.

I'll have to think about this more - created #19 to remind myself.
